Add mmap-based fallback solution if CMA is not available

zhngaj opened the pull request:
This series of commit includes two parts
1. Added mmap-based fallback solution in shm provider if CMA is not available
2. Made corresponding change in efa provider

Note:
1. This work can be rebased onto PR #5664, once it's merged. 
2. Open this PR first for community review.
3. The fabtests for shm provider (especially fi_ubertest) takes longer time. We might need to increase the timeout value (`-N 1200` works fine on my end).

zhngaj commented:
AWS CI failed with CI system related failure.

aingerson commented:
I would suggest a different approach. This is adding a lot of code and fields to the more generic paths (like adding the map_ptr and map_name to the smr_resp). I think it would be best to add another smr_src_* to use this protocol. It's different enough from smr_src_iov to separate it.

I'm also confused on the implementation and usage of the peer_msg_id. I don't understand why this is necessary. It seems to me like this could be included in the smr_cmd if we outline a new protocol for this.

I think the CMA check should also be moved outside of the message path and done on a per peer basis during shm mapping. Right now, the check is happening while a message is trying to send which adds a lot of code to the message path (even though most of it is only done once, it's still unnecessary to include there when it could be done elsewhere). I'm not completely sure, but I think this could also be problematic if peers have different abilities for CMA. I have a patch that implements this check during smr_map_to_endpoint and saves the CMA capability per peer. That way, if doing a CMA from peer 0 to peer 1 is allowed but not to peer 2, we don't have issues with sending to peer 2 if we already decided to allow CMA copies to any peer from peer 0. Again, I could be incorrect about this in which case there's no reason to do it per peer but we could still move it to the initialization/mapping code.

Lastly, I want to clarify what your understanding is so we're on the same page. Are you imagining the shm mapping protocol to replace the large message segmentation or be an additional protocol? Asking because this affects the PR I have open right now regarding reading without CMA.

shefty commented:
We should probably open a discussion up on the mailing list to gather input on the best approaches to shared memory transfers.  mmap may require very large transfers to outperform bounce buffer approaches.  And bounce buffers may need multiple threads involved in the copies to overlap copy-in with copy-out.  I have not analyzed the MPI shared memory implementations to see how they have handled this.

zhngaj commented:
> I would suggest a different approach. This is adding a lot of code and fields to the more generic paths (like adding the map_ptr and map_name to the smr_resp). I think it would be best to add another smr_src_* to use this protocol. It's different enough from smr_src_iov to separate it.

Sounds reasonable. I can add a separate protocol, say `smr_src_mmap`.  

> 
> I'm also confused on the implementation and usage of the peer_msg_id. I don't understand why this is necessary. It seems to me like this could be included in the smr_cmd if we outline a new protocol for this.

`msg_id` is used to differentiate the long message transfers between the same two endpoints.  An EP keeps track of the msg_id it sends to different peers individually. I feel this is intuitive. Could you elaborate a bit about including it in the smr_cmd in new protocol? 

> 
> I think the CMA check should also be moved outside of the message path and done on a per peer basis during shm mapping. Right now, the check is happening while a message is trying to send which adds a lot of code to the message path (even though most of it is only done once, it's still unnecessary to include there when it could be done elsewhere). I'm not completely sure, but I think this could also be problematic if peers have different abilities for CMA. I have a patch that implements this check during smr_map_to_endpoint and saves the CMA capability per peer. That way, if doing a CMA from peer 0 to peer 1 is allowed but not to peer 2, we don't have issues with sending to peer 2 if we already decided to allow CMA copies to any peer from peer 0. Again, I could be incorrect about this in which case there's no reason to do it per peer but we could still move it to the initialization/mapping code.

We had a discussion here https://github.com/ofiwg/libfabric/pull/5510#issuecomment-567708774, where we think the CMA check could be delayed till the real communication starts. But moving it to initialization/mapping code is straightforward. I can do it. 
I don't think we need to consider the per peer CMA capability, as the ptrace_scope level is set across the system. All local peers should have the consistent/same check result. 

> 
> Lastly, I want to clarify what your understanding is so we're on the same page. Are you imagining the shm mapping protocol to replace the large message segmentation or be an additional protocol? Asking because this affects the PR I have open right now regarding reading without CMA.

Our idea is that SAR solution will be the default fallback solution, and mmap-based fallback will be considered as an alternative solution. Depending on the performance tuning results, we can have/set a threshold to decide when to switch from SAR to mmap. Please let me know if this sounds reasonable.

aingerson commented:
> `msg_id` is used to differentiate the long message transfers between the same two endpoints. An EP keeps track of the msg_id it sends to different peers individually. I feel this is intuitive. Could you elaborate a bit about including it in the smr_cmd in new protocol?
If each endpoint has a unique shm prefix (like the pid:domain_index:endpoint_index combination that we have or the shm name which is already saved in the smr_region), then the only thing we need is an msg_id per ep to create a unique shm name for that transfer. Then the sender/mapper would just need to send the msg_id and the receiver could just open the shm region called <ep_prefix>:cmd.msg.hdr.msg_id

aingerson commented:
> We had a discussion here [#5510 (comment)](https://github.com/ofiwg/libfabric/pull/5510#issuecomment-567708774), where we think the CMA check could be delayed till the real communication starts. But moving it to initialization/mapping code is straightforward. I can do it.
> I don't think we need to consider the per peer CMA capability, as the ptrace_scope level is set across the system. All local peers should have the consistent/same check result.

Yeah I don't think that it would be on your systems, but I'm thinking to other situations. I'm not sure if this is a realistic situation though. I just implemented it that way as a precaution that it might be. Maybe @shefty can provide insight on this?

aingerson commented:
> Our idea is that SAR solution will be the default fallback solution, and mmap-based fallback will be considered as an alternative solution. Depending on the performance tuning results, we can have/set a threshold to decide when to switch from SAR to mmap. Please let me know if this sounds reasonable.

Yes, this seems reasonable to me. We should coordinate more then about what updates and changes we are making. The PR I have open is using a similar method to how I'm implementing the segmentation protocol and it includes the CMA checking as well as some refactoring of the usage of msg_id/pending_cmd. I'll try to push those changes soon so you can take a look at what I'm thinking.

shefty commented:
I doubt we'll have per peer ptrace support in practice.  But that's a guess.

zhngaj commented:
Based on mmap fallback's basic workflow that we have agreed with earlier [mmap-based-fallback-workflow.docx](https://github.com/ofiwg/libfabric/files/4283882/mmap-based-fallback-workflow.docx) and the above comments, I want to summary the next steps that I plan to proceed with.

![image](https://user-images.githubusercontent.com/48572887/75835395-a0e1db80-5d73-11ea-98fd-998f6567526f.png)

@aingerson Can you please take a look to see if it looks good to you?

aingerson commented:
@zhngaj Yes, this looks like a good proposal. And the smr_tx_entry exactly matches the entry that I have in my patch so that is great :)
My only suggestion would be to remove the ep_name from the tx_entry since we should be able to build that from the ep/region prefix + cmd.msg.hdr.msg_id

zhngaj commented:
@aingerson, thanks for your feedback.

> My only suggestion would be to remove the ep_name from the tx_entry since we should be able to build that from the ep/region prefix + cmd.msg.hdr.msg_id

I'm assuming you were referring to `struct smr_ep_name *map_name`. The reason I added it is that we need the entry, which is used to insert into the existing `ep_name_list`, 
dlist_insert_tail(&(*map_name)->entry, &ep_name_list);
This way we can delete all stale shm files in case of an unclean shutdown via shm's signal handling logic. Please let me know if it sounds reasonable.

aingerson commented:
@zhngaj If it has already been inserted into the ep_name_list for cleanup later, why does it need to be stored in the tx_entry?

zhngaj commented:
I was thinking to remove the entry from the `ep_name_list` if it's a clean shutdown upon receiving the response. 
I see your point. If it's a clean shutdown, `ep_name_list` will be cleanedup via smr_cleanup.

aingerson commented:
Hmm, yeah that makes more sense. Doesn't make sense to no clean these names up as the transfers are done. That's fine then. Thanks for the explanation.

zhngaj commented:
@aingerson Just to be clear, so you are okay with keeping `struct smr_ep_name *map_name` in the `smr_tx_entry`.

aingerson commented:
@zhngaj Yes, I think that's ok for now. I don't know if there's a way to easily do this, but ideally I think storing the ep_names per endpoint instead of per process would be better so the searching and removing doesn't get very expensive.

zhngaj commented:
@aingerson @shefty @ofiwg/aws-efa-team Can you please review the updated change? 
One thing I want to callout is that I moved the CMA check to smr_map_to_endpoint instead of smr_map_to_region in the proposal (item 2). 
Item 1 and 3 are following the proposal exactly.

aingerson commented on prov/shm/src/smr_ep.c:
```diff
@@ -303,17 +308,106 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+	cmd->msg.hdr.src_data = (uintptr_t) ((char **) resp - (char **) smr);
+	cmd->msg.data.iov_count = count;
+	cmd->msg.hdr.size = total_len;
+	memcpy(cmd->msg.data.iov, iov, sizeof(*iov) * count);
```
I don't think this is necessary anymore since we're saving this directly in the tx_entry struct now

aingerson commented on prov/util/src/util_shm.c:
```diff
@@ -62,6 +62,25 @@ static void smr_peer_addr_init(struct smr_addr *peer)
+		peer_region->cma_cap = 1;
+	}
+}
+
```
I don't see the need to store the remote_iov here or writing to the pid.
We should just be able to write directly to the peer_smr->cma_cap and use a simple address translation if we just save the smr base address.
ie:
Initialize
```
smr->cma_cap = -1;
smr->base_addr = *smr;
```
And then here do something like this
```
smr_cma_check(...)
{
    local_iov.iov_base = &smr->cma_cap;
    remote_iov.iov_base = peer_smr->base_addr + (peer_smr->cma_cap - peer_smr);
    process_vm_writev(peer_smr->pid, &local_iov, 1, &remove_iov, 1);
}
```

zhngaj replied:
Good point. Fixed in the new revision. Thank you.

zhngaj commented on prov/shm/src/smr_ep.c:
```diff
@@ -303,17 +308,106 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+	cmd->msg.hdr.src_data = (uintptr_t) ((char **) resp - (char **) smr);
+	cmd->msg.data.iov_count = count;
+	cmd->msg.hdr.size = total_len;
+	memcpy(cmd->msg.data.iov, iov, sizeof(*iov) * count);
```
Thank you. Fixed in the new revision.

zhngaj replied:
Thank you. Fixed in new revision. 
I see an issue when doing the same change in the format_iov though, so I didn't change in format_iov. I will do more testing and fix it in later revision or via another PR.

zhngaj replied:
We cannot remove the memcpy in format_iov, as we cannot access the tx_entry struct via resp->msg_id in the peer's memory space.

zhngaj commented on prov/shm/src/smr_ep.c:
```diff
@@ -303,17 +308,106 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+	pend->map_name = map_name;
+	if (cmd->msg.hdr.op == ofi_op_read_req) {
+		pend->map_ptr = shm_ptr;
+	}
```
Make sense. Thanks.

zhngaj commented on prov/shm/src/smr_msg.c:
```diff
@@ -212,9 +215,31 @@ static ssize_t smr_generic_sendmsg(struct smr_ep *ep, const struct iovec *iov,
+			ret = smr_iov_mmap_copy_in(ep, peer_smr, iov, iov_count,
+						   total_len, op, msg_id, &map_name, NULL);
+			if (ret) {
+				ofi_cirque_discard(smr_resp_queue(ep->region));
```
You are right. Thanks.

zhngaj commented on prov/shm/src/smr_rma.c:
```diff
@@ -159,9 +163,31 @@ ssize_t smr_generic_rma(struct smr_ep *ep, const struct iovec *iov,
+						   iov_count, total_len, op, msg_id,
+						   &map_name, &shm_ptr);
+			if (ret) {
+				ofi_cirque_discard(smr_resp_queue(ep->region));
```
Thanks.

zhngaj commented:
AWS CI failed with CI system related failure (Ubuntu with tcp provider, ssh to slave instance timed out).
https://libfabric-ci.aws.a2z.com/job/multi-node/4/PROVIDER=tcp,label=ubuntu/console

rajachan commented on prov/efa/src/rxr/rxr.h:
```diff
@@ -825,7 +826,7 @@ int rxr_ep_set_tx_credit_request(struct rxr_ep *rxr_ep,
 
 void rxr_inline_mr_reg(struct rxr_domain *rxr_domain,
-		       struct rxr_tx_entry *tx_entry);
+		       struct rxr_tx_entry *tx_entry, bool is_shm);
```
Can we not tell if this is a shm transfer from the tx_entry or the peer struct? Do we need this signature change?

zhngaj replied:
Then inside rxr_inline_mr_reg, we can do  
```
peer = rxr_ep_get_peer(ep, tx_entry->addr);
if (peer->is_local) {}
```
Is this acceptable to you?

rajachan replied:
Yep, that's what I meant.

zhngaj replied:
I was thinking to use peer struct too, but it requires the rxr_ep, which is not available here. Even we have mr squash #5664 merged, I think we still need this signature change, as we need to check `if (tx_entry->iov[index].iov_len > rxr_env.max_memcpy_size)` for efa only.

wzamazon replied:
We can replace rxr_domain with rxr_ep

zhngaj replied:
If replacing rxr_domain with rxr_ep sounds good to you @rajachan, I can update.

rajachan replied:
Registration is generally a domain-level construct, hence the original signature, but you can get to the domain from the EP so that's fine.

rajachan commented on prov/efa/src/rxr/rxr_ep.c:
```diff
@@ -506,12 +506,26 @@ struct rxr_tx_entry *rxr_ep_alloc_tx_entry(struct rxr_ep *rxr_ep,
+			ret = fi_mr_reg(&rxr_domain->util_domain.domain_fid,
+					tx_entry->iov[index].iov_base,
+					tx_entry->iov[index].iov_len,
+					FI_REMOTE_READ, 0, 0, 0,
```
Is this the only access you'd need for this registration?

zhngaj replied:
Yes. It's only used for the peer's RMA Read operation.

wzamazon replied:
Can you reuse the registration in this function?

zhngaj replied:
Do you mean reuse the following function and pass different arguments for struct *fid_domain and int access?
```
545                         ret = fi_mr_reg(rxr_domain->rdm_domain,
546                                         tx_entry->iov[index].iov_base,
547                                         tx_entry->iov[index].iov_len,
548                                         FI_SEND, 0, 0, 0,
549                                         &tx_entry->mr[index], NULL);

```

zhngaj replied:
Function rxr_inline_mr_reg is re-formatted to fit for shm and efa cases.

rajachan commented on prov/efa/src/rxr/rxr_ep.c:
```diff
@@ -506,12 +506,26 @@ struct rxr_tx_entry *rxr_ep_alloc_tx_entry(struct rxr_ep *rxr_ep,
+	if (is_shm) {
+		tx_entry->iov_mr_start = index;
+		while (index < tx_entry->iov_count) {
+			ret = fi_mr_reg(&rxr_domain->util_domain.domain_fid,
```
I remember our conversation about using a regv() instead, but someone has to have this loop given our mr_Iov_limit, either here or the bottom layer. I wonder if https://github.com/ofiwg/libfabric/pull/5664 makes this better from a code perspective (have not read that PR yet).

zhngaj replied:
#5664 right now does not handle the loop. It does ibv_reg_mr, ofi_mr_map_insert, and shm 's fi_mr_reg. 
Since we already had this loop for rdm domain, I just keep the same pattern here. If we decide to lower this loop down to efa_mr_reg*, I will change accordingly. @dipti-kothari @wzamazon please let me know your thoughts.

rajachan replied:
Nah, that's alright, I was just curious. After reading the other PR, I think we can leave this be for now and revist how we do regv after all the squash changes land.

zhngaj replied:
Sure.

rajachan commented on prov/efa/src/rxr/rxr_pkt_type_misc.c:
```diff
@@ -490,11 +490,23 @@ void rxr_pkt_handle_eor_recv(struct rxr_ep *ep,
+		for (i = 0; i < tx_entry->rma_iov_count; i++) {
+			if (tx_entry->mr[i]) {
+				ret = fi_close(&tx_entry->mr[i]->fid);
+				if (ret) {
```
OFI_UNLIKELY

zhngaj replied:
Fixed

wzamazon replied:
These has been done in https://github.com/ofiwg/libfabric/pull/5721 the 3rd commit. We need to compare notes on this

zhngaj replied:
Updated with the one in #5721.

rajachan commented on prov/efa/src/rxr/rxr_pkt_type_misc.c:
```diff
@@ -490,11 +490,23 @@ void rxr_pkt_handle_eor_recv(struct rxr_ep *ep,
 	struct rxr_tx_entry *tx_entry;
+	int i, ret;
 
 	shm_eor = (struct rxr_eor_hdr *)pkt_entry->pkt;
```
Isn't `rxr_pkt_handle_eor_recv`used for both the shm and non-shm paths? How is this directly treating it as a shm eor packet?

zhngaj replied:
Your are right, thank you, will fix it.

wzamazon commented on prov/efa/src/rxr/rxr.h:
```diff
@@ -188,6 +188,7 @@ static inline void rxr_poison_mem_region(uint32_t *ptr, size_t size)
 #define RXR_SHM_HDR_DATA	BIT_ULL(11)
 
 extern struct fi_info *shm_info;
+extern int efa_cma_cap;
```
Can we move this variable to inside rxr_env? also, shouldn't it named shm_cma_cap because EFA does not use CMA?

zhngaj replied:
Sure, will name it as shm_cma_cap.

wzamazon commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -155,7 +155,13 @@ struct rxr_read_entry *rxr_read_alloc_entry(struct rxr_ep *ep, int entry_type, v
+		 */
+		for (i = 0; i < read_entry->rma_iov_count; i++)
+			read_entry->rma_iov[i].addr = (efa_cma_cap) ?
+						      read_entry->rma_iov[i].addr : 0;
```
Why not
```
if (!efa_cma_cap) {
    for (I=0; I < read_entry->rma_iov_count; ++i)
       read_entry->rma_iov[i].addr = 0;
}
```
and the setting of rma_iov[I] should be earlier. This place is for setting local mr. 

you can register local mr here instead of in ```rxr_inline_mr_reg```

zhngaj replied:
1. I can do the rma_iov[i] setting in function rxr_pkt_proc_shm_long_msg_rts() before calling rxr_read_alloc_entry(). 
2. For long message via shm, we need to do iov registration on the initiator side (initiator's registered `iov` will be sent to remote as `rma_iov`). I don't think the logic I added in rxr_inline_mr_reg can be done here (remote side). Please let me know if I miss something.

wzamazon commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -263,6 +269,7 @@ int rxr_read_post(struct rxr_ep *ep, struct rxr_read_entry *read_entry)
 		 */
 		if (read_entry->lower_ep_type == SHM_EP)
 			pkt_entry = rxr_pkt_entry_alloc(ep, ep->tx_pkt_shm_pool);
+
```
Nit, unnecessary blank line

zhngaj replied:
Will fix.

wzamazon commented on prov/efa/src/rxr/rxr_rma.c:
```diff
@@ -171,6 +172,14 @@ size_t rxr_rma_post_shm_write(struct rxr_ep *rxr_ep, struct rxr_tx_entry *tx_ent
+	for (i = 0; i < tx_entry->rma_iov_count; i++) {
+		tx_entry->rma_iov[i].addr = (efa_cma_cap) ?
+					    tx_entry->rma_iov[i].addr : 0;
+	}
```
same comments apply here.

zhngaj replied:
I will use format
```
if (!efa_cma_cap) {
    for (I=0; I < read_entry->rma_iov_count; ++i)
       read_entry->rma_iov[i].addr = 0;
}
```

zhngaj commented:
Thanks for the reminder. I think the shm long message logic in current RXR_RTS_PKT can be combined with RXR_LONG_TAGRTM_PK, RXR_LONG_MSGRTM_PKT later.

aingerson commented on prov/shm/src/smr_rma.c:
```diff
@@ -142,7 +146,7 @@ ssize_t smr_generic_rma(struct smr_ep *ep, const struct iovec *iov,
 	cmd = ofi_cirque_tail(smr_cmd_queue(peer_smr));
 
-	if (cmds == 1) {
+	if (cmds == 1 && ep->region->cma_cap) {
```
The cma_cap check should get added on line 130 where we calculate whether to go to the fast path or not

zhngaj replied:
Done.

aingerson commented on prov/shm/src/smr_rma.c:
```diff
@@ -159,9 +163,30 @@ ssize_t smr_generic_rma(struct smr_ep *ep, const struct iovec *iov,
+			smr_format_mmap(cmd, smr_peer_addr(ep->region)[peer_id].addr,
+					iov, iov_count, total_len, op, 0, data,
+					op_flags, context, ep->region, msg_id,
+					map_name, shm_ptr, resp, pend);
```
Would it be possible to combine mmap_copy_in and format_mmap? We're passing in the same arguments and some of them could then be removed from the declaration in this function (and in the generic_sendmsg function)

zhngaj replied:
I combined mmap_copy_in with format_mmap. Thank you.

aingerson commented on prov/util/src/util_shm.c:
```diff
@@ -62,6 +62,22 @@ static void smr_peer_addr_init(struct smr_addr *peer)
+{
+	struct iovec local_iov, remote_iov;
+	int ret;
+
```
We could also add a check here to see if the peer has already checked this. ie
```
   if (peer_smr->cma_cap != -1) {
        smr->cma_cap = peer_smr->cma_cap;
        return;
   }
```

zhngaj replied:
Thanks for your suggestion. Updated.

aingerson commented on prov/shm/src/smr_progress.c:
```diff
@@ -202,8 +217,93 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
+	if (mapped_ptr == MAP_FAILED) {
+		FI_WARN(&smr_prov, FI_LOG_AV, "mmap error %s\n", strerror(errno));
+		return errno;
+	}
```
On failure here, do we still need to shm_unlink?

zhngaj replied:
Yeah, updated, thanks.

shefty replied:
-errno and cleanup

zhngaj replied:
Done.

> > + if (mr_mode & FI_MR_VIRT_ADDR) + cur->domain_attr->mr_mode = FI_MR_VIRT_ADDR; My understanding is that if app requires FI_MR_VIRT_ADDR via hints, and shm provider supports FI_MR_VIRT_ADDR, the returned info should have FI_MR_VIRT_ADDR bit, regardless of CMA capability.
> FI_MR_VIRT_ADDR is a requirement *from the provider*, not the app. The app indicates if it can support that requirement. The provider toggles the bit on/off if it is needed.

Sorry for the confusion. Let me put my question in this way:
suppose that app puts FI_MR_VIRT_ADDR bit in hints, and the app calls fi_getinfo. 
Is FI_MR_VIRT_ADDR bit supposed to be returned from shm provider?

The current logic (without my change) is: 
- if CMA available, rma fast path will be enabled, FI_MR_VIRT_ADDR will be set and returned.
- if no CMA, rma fast path will be disabled, thus NO FI_MR_VIRT_ADDR bit will be set and retuned.
Is such behavior expected?

aingerson commented on prov/shm/src/smr_progress.c:
```diff
@@ -202,8 +217,93 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
+		if (ret != cmd->msg.hdr.size) {
+			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+				"mmap iov copy out error\n");
+			return FI_EIO;
```
On error here, do we need to unlink and unmap?

zhngaj replied:
Thank you. Fixed.

shefty replied:
error return codes should be negative for consistency with the rest of the code base.

shefty replied:
We need to fix the function name.  Are we copying into the mmap region or out of the mmap region, and into or out of the iov?

zhngaj replied:
For send-recv, and RMA write, it's mmap region->iov, while for RMA read, it's iov->mmap region. So I'm not sure what the good function name would be. Do you have any suggestion?

zhngaj replied:
I renamed it to smr_mmap_peer_copy(), please let me know if it looks good to you.

aingerson commented on prov/shm/src/smr_rma.c:
```diff
@@ -159,9 +163,30 @@ ssize_t smr_generic_rma(struct smr_ep *ep, const struct iovec *iov,
+			 * Once SAR protocol gets merged, we need set a threshold
+			 * for switching from SAR to mmap
+			 */
+			msg_id = (ep->msg_id)++;
```
We declare and pass this in? Can we move this increment into the format call?

zhngaj replied:
Yeah. I moved this into the combined format_mmap call.

zhngaj commented:
> This patch seems to be relying on RTS, but we will retire RTS eventually. please keep that in mind

I have removed the changes to RTS for efa read message protocol via shm, and moved it to rxr_read_init_iov.

shefty commented on prov/shm/src/smr_ep.c:
```diff
@@ -302,17 +307,102 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+	int fd, ret, num;
+
+	sprintf(peer_name, "%s", (char *)peer_smr + peer_smr->name_offset);
+	num = sprintf(shm_name, "%s_%s_%ld", ep->name, peer_name, msg_id);
```
Please use snprintf, so we don't immediately need to respond to static code analysis complaining about possible overflow.

zhngaj replied:
Sure. Updated with snprintf.

zhngaj commented:
@ofiwg/aws-efa-team @aingerson @shefty 
Thanks for your comments. I addressed all of them and updated the PR. Please let me know if you have further comments.

wzamazon commented on prov/efa/src/rxr/rxr_pkt_type_rts.c:
```diff
@@ -223,32 +223,12 @@ ssize_t rxr_pkt_init_rts(struct rxr_ep *ep,
+	data_len = ofi_copy_from_iov(data, mtu - pkt_entry->pkt_size,
+				     tx_entry->iov, tx_entry->iov_count, 0);
+	assert(data_len == rxr_get_rts_data_size(ep, rts_hdr));
+	pkt_entry->pkt_size += data_len;
```
RTS is not used any more. Any change to this file is unnecessary.

zhngaj replied:
I only remove the shm related logic here, and didn't touch the efa part. If we want to remove the whole RTS code, I prefer doing it in a separate PR. Does it acceptable to you?

zhngaj replied:
@wzamazon meant to not change anything, just leave as what it was. Updated.

wzamazon commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -230,12 +237,14 @@ int rxr_read_init_iov(struct rxr_ep *ep,
-			assert(!tx_entry->mr[i]);
 			read_iov[i].addr = (uint64_t)tx_entry->iov[i].iov_base;
 			read_iov[i].len = tx_entry->iov[i].iov_len;
-			read_iov[i].key = 0;
```
I do not think you need call rxr_inline_mr_reg() because this function also does memory registration will you can use.

I think you just need to adjust the condition a little bit to:

```
if (tx_entry->desc[I]) {
    // no registration 
} else if (shm && rxr_env.shm_cma_cap) {
    rma_iov[I].key = 0
} else {
    // memory registration and set rma_iov[I].key.
}
```

zhngaj replied:
Thank you. Updated.

wzamazon commented on prov/efa/src/rxr/rxr_rma.c:
```diff
@@ -171,11 +173,26 @@ size_t rxr_rma_post_shm_write(struct rxr_ep *rxr_ep, struct rxr_tx_entry *tx_ent
 	msg.context = pkt_entry;
 	msg.data = tx_entry->cq_entry.data;
-	return fi_writemsg(rxr_ep->shm_ep, &msg, tx_entry->fi_flags);
+	ret = fi_writemsg(rxr_ep->shm_ep, &msg, tx_entry->fi_flags);
```
We need a discussion here. If shm decided to return -FI_EAGIN, why do we want to queue them?

zhngaj replied:
Team reaches an agreement to not add the queuing logic. The current queuing for read will be revisited later.

wzamazon commented on prov/efa/src/rxr/rxr_rma.h:
```diff
@@ -47,7 +47,7 @@ struct rxr_tx_entry *
 			       struct rxr_rx_entry *rx_entry);
 
-size_t rxr_rma_post_shm_rma(struct rxr_ep *rxr_ep,
+size_t rxr_rma_post_shm_write(struct rxr_ep *rxr_ep,
```
I think this declaration can be deleted because the function ```rxr_rma_post_shm_write``` was only used inside ```rxr_rma.c```

zhngaj replied:
Yes, thanks.

wzamazon commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -155,7 +155,14 @@ struct rxr_read_entry *rxr_read_alloc_entry(struct rxr_ep *ep, int entry_type, v
+		if (!rxr_env.shm_cma_cap) {
+			for (i = 0; i < read_entry->rma_iov_count; i++)
+				read_entry->rma_iov[i].addr = 0;
+		}
```
please move the change of rma_iov to the next part, and the line of ```memset``` should not be removed.

zhngaj replied:
@wzamazon and I think it's better to let shm expose FI_MR_VIR_ADDR bit regardless of CMA capability of system, so we don't need to deal with 0-based offset.

zhngaj commented on prov/efa/src/rxr/rxr_init.c:
```diff
@@ -74,6 +75,7 @@ struct rxr_env rxr_env = {
 
 static void rxr_init_env(void)
 {
+	fi_param_get_int(&rxr_prov, "shm_cma_cap", &rxr_env.shm_cma_cap);
```
Sure. I can remove it.

zhngaj commented on prov/efa/src/rxr/rxr_init.c:
```diff
@@ -626,6 +628,8 @@ struct fi_provider rxr_prov = {
 
 EFA_INI
 {
+	fi_param_define(&rxr_prov, "shm_cma_cap", FI_PARAM_INT,
```
Will do.

shefty commented on prov/shm/src/smr_msg.c:
```diff
@@ -192,9 +194,24 @@ static ssize_t smr_generic_sendmsg(struct smr_ep *ep, const struct iovec *iov,
+					iov_count, total_len, op, tag, data, op_flags,
+					context, ep->region, msg_id, map_name, resp, pend);
+
+		}
```
I'd like to out-dent this and maybe find a way to re-work the if - else if - else protocol selection, so we can prioritize the small transfers and make it clearer that's the purpose of the checks.  As the code is, we have a series of checks and jumps before we can do inline transfers (smr_format_inline), and the code doesn't clearly convey that we're really changing the protocol.

zhngaj replied:
I re-ordered to protocol selection, and prioritize inline/inject transfer.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -104,6 +104,15 @@ static void smr_progress_resp(struct smr_ep *ep)
 		}
+
+		if (!ep->region->cma_cap) {
+			if (pending->map_name) {
```
Isn't the check for map_name sufficient?  Would that ever be set if we weren't using mmap?  It's a more direct check then !cma_cap, which just means we're not using cma.

zhngaj replied:
Yeah. Updated, thank you.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -200,6 +209,78 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
+	peer_id = (int) cmd->msg.hdr.addr;
+
+	sprintf(peer_name, "%s", ep->region->map->peers[peer_id].peer.name);
+	num = sprintf(shm_name, "%s_%s_%ld", peer_name, ep->name,
```
need snprintf here

zhngaj replied:
Done.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -200,6 +209,78 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
+		      cmd->msg.hdr.msg_id);
+	if (num < 0) {
+		FI_WARN(&smr_prov, FI_LOG_AV, "generating shm file name failed\n");
+		return errno;
```
Negative errno would be consistent with the rest of the code base.

zhngaj replied:
Done.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -200,6 +209,78 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
+	fd = shm_open(shm_name, O_RDWR, S_IRUSR | S_IWUSR);
+	if (fd < 0) {
+		FI_WARN(&smr_prov, FI_LOG_AV, "shm_open error\n");
+		return errno;
```
- errno and need to cleanup on error

zhngaj replied:
Done.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -200,6 +209,78 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
+	if (ret != cmd->msg.hdr.size) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"mmap iov copy out error\n");
+		return FI_EIO;
```
- EIO and cleanup

zhngaj replied:
Done.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -200,6 +209,78 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
+
+	if (err) {
+		ret = -err;
+		goto out;
```
This is odd behavior. The caller passes in err, which if it's non-zero, we just return negated?

zhngaj replied:
I was following the pattern of smr_progress_inject and smr_progress_iov. But I agree with you. This snippet is useless. I've removed it.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -200,6 +209,78 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
+
+	peer_id = (int) cmd->msg.hdr.addr;
+	peer_smr = smr_peer_region(ep->region, peer_id);
+	resp = (struct smr_resp *) ((char **) peer_smr +
```
why char **?  Is there a macro for this?

Why are we even getting peer_smr and resp?  These are local variables that we set and then lose on function exit?  Something is off here.

zhngaj replied:
I went through my PR to use char * in such case, and added macros for address translation.

This is the at receiver/target side, and it only receives `cmd`. 
resp is still at sender/initiator side. It's not sent to the receiver/target side. Therefore we need to find the resp struct on the sender/initiator side.

shefty commented on prov/shm/src/smr_ep.c:
```diff
@@ -381,14 +382,17 @@ int smr_iov_mmap_copy_in(struct smr_ep *ep, struct smr_region *peer_smr,
+		}
+		munmap(mapped_ptr, total_len);
+	} else {
+		*shm_ptr = mapped_ptr;
```
This change makes the function confusing and the encapsulation weird.  The function is called 'copy_in', but now no copy is being performed, and the only use is to set a pointer using a value passed in by the caller.  So, the caller could just have easily set this.  And now the mmap is left, but the file is closed.

zhngaj replied:
We have combined copy_in() and format_mmap(). So shm file is opened and mmap'ed for all cases. 

> And now the mmap is left, but the file is closed.
I think it's valid.
As per http://man7.org/linux/man-pages/man2/mmap.2.html. 
After the mmap() call has returned, the file descriptor, fd, can be closed immediately without invalidating the mapping.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -241,16 +249,29 @@ static int smr_mmap_iov_copy_out(struct smr_ep *ep, struct smr_cmd *cmd,
 	munmap(mapped_ptr, cmd->msg.hdr.size);
-	close(fd);
 	shm_unlink(shm_name);
+	close(fd);
```
This should be fixed in the patch that introduces the change, or come as a separate fix

zhngaj replied:
Done.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -241,16 +249,29 @@ static int smr_mmap_iov_copy_out(struct smr_ep *ep, struct smr_cmd *cmd,
+				"mmap iov copy out error\n");
+			return FI_EIO;
+		}
+		*total_len = ret;
```
This isn't needed, as total_len is set 2 lines down.

zhngaj replied:
I removed the setting 2 lines down.

shefty commented on prov/shm/src/smr_rma.c:
```diff
@@ -160,9 +163,30 @@ ssize_t smr_generic_rma(struct smr_ep *ep, const struct iovec *iov,
+						   iov_count, total_len, op, msg_id,
+						   &map_name, &shm_ptr);
+			if (ret) {
+				freestack_push(ep->pend_fs, pend);
```
Why is the stack push not part of the common error path?

zhngaj replied:
This is because only format_mmap (combined copy_in and original format_mmap) call could fail. format_iov call just does bunch of assignments.

shefty commented on prov/shm/src/smr_ep.c:
```diff
@@ -386,23 +371,34 @@ int smr_iov_mmap_copy_in(struct smr_ep *ep, struct smr_region *peer_smr,
 	close(fd);
-err1:
-	return -errno;
+	return ret;
```
Please don't have patches modify other patches in the series.  Update the original patch directly.

zhngaj replied:
Updated.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -222,20 +222,18 @@ static int smr_mmap_iov_copy_out(struct smr_ep *ep, struct smr_cmd *cmd,
-		      cmd->msg.hdr.msg_id);
+	num = snprintf(shm_name, NAME_MAX - 1, "%s_%s_%ld",
+		     ep->region->map->peers[peer_id].peer.name, ep->name,
+		     cmd->msg.hdr.msg_id);
```
Don't mix changes together in a single patch.

zhngaj replied:
Updated

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -247,33 +245,38 @@ static int smr_mmap_iov_copy_out(struct smr_ep *ep, struct smr_cmd *cmd,
 				"mmap iov copy out error\n");
-			return FI_EIO;
+			ret = FI_EIO;
+			goto err2;
```
Update original patch.  Don't combine unrelated changes together.

zhngaj replied:
Done.

shefty commented on prov/shm/src/smr_ep.c:
```diff
@@ -319,56 +319,41 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+		    const struct iovec *iov, size_t count, size_t total_len,
+		    uint32_t op, uint64_t tag, uint64_t data, uint64_t op_flags,
+		    void *context, struct smr_resp *resp,
+		    struct smr_tx_entry *pend)
```
Please rename, as this is no longer just formatting.

zhngaj replied:
I renamed it to smr_mmap_copy_format(). Please let me know if the new name looks okay or not.

wzamazon commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -264,8 +264,6 @@ int rxr_read_init_iov(struct rxr_ep *ep,
 		for (i = 0; i < tx_entry->iov_count; ++i) {
 			assert(tx_entry->mr[i]);
-			read_iov[i].addr = (uint64_t)tx_entry->iov[i].iov_base;
-			read_iov[i].len = tx_entry->iov[i].iov_len;
```
Let's do memory registration for shm regardless shm_cma_cap, eg:

```
if (tx_entry->desc[0]) {
     read_iov[I].key = fi_rma_key(desc)
} else{
    memory registration
    read_iov[I].key = fi_rma_key(mr[I])
}
```
Because this is the correct way to do things, and also get ride of shm_cma_cap altogether.

zhngaj replied:
But that would affect performance as we are doing unnecessary mr registration.

wzamazon replied:
With MR cache, I do not think it will have big impact. 

But your point is also valid, let us keep it this way, and run some tests to determine if we can do what I suggested.

zhngaj commented on prov/efa/src/rxr/rxr_init.c:
```diff
@@ -145,7 +145,7 @@ int rxr_ep_efa_addr_to_str(const void *addr, char *smr_name)
 	qpn = ((struct efa_ep_addr *)addr)->qpn;
 
-	ret = snprintf(smr_name, NAME_MAX, "/%ld_%s_%d", (size_t) getuid(), gid, qpn);
+	ret = snprintf(smr_name, NAME_MAX, "%ld_%s_%d", (size_t) getuid(), gid, qpn);
```
Yes, it needs.

zhngaj commented on prov/util/src/util_shm.c:
```diff
@@ -62,6 +62,26 @@ static void smr_peer_addr_init(struct smr_addr *peer)
+	struct iovec local_iov, remote_iov;
+	int ret;
+
+	if (peer_smr->cma_cap != -1) {
```
It can be. As we need to execute a process_vm_writev call to try to write on the peer's cma_cap address, in the latest revision I set cma_cap as a separate flag which is different from the current `uint16_t flags`, so that we don't mess up the value in current flags. Is it okay?

zhngaj commented on prov/shm/src/smr.h:
```diff
@@ -234,6 +236,12 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+		uint32_t op, uint64_t tag, uint64_t data, uint64_t op_flags,
+		void *context, struct smr_region *smr, uint64_t msg_id,
+		struct smr_ep_name *map_name,
+		struct smr_resp *resp, struct smr_tx_entry *pend);
```
Yeah. I'm thinking a way to clean it up and I will open a separate PR to do it on all shm functions.

aingerson replied:
By the way, I have a patch ready to clean it up that I'm planning to push to my open PR later today once I finalize the other changes. So maybe hold off on that so we're using the same changes and don't create a lot of conflicts for each other.

zhngaj replied:
Sure.

zhngaj commented on prov/shm/src/smr_ep.c:
```diff
@@ -319,6 +319,88 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+
+	smr_post_pend_resp(pend, cmd, context, iov, count, resp);
+
+	pend->map_name = map_name;
```
No particular reason. The map_name was being saved at the end of post_pend_resp function before. Since we combined mmap_copy_in() and format_mmap(), I just put it after post_pend_resp call. 
I have moved it before pos_pend_resp call.

zhngaj commented on prov/shm/src/smr_ep.c:
```diff
@@ -319,6 +319,88 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+	smr_generic_format(cmd, peer_id, op, tag, 0, 0, data, op_flags);
+	cmd->msg.hdr.op_src = smr_src_mmap;
+	cmd->msg.hdr.msg_id = msg_id;
+	cmd->msg.hdr.src_data = (uintptr_t) ((char **) resp - (char **) smr);
```
I was just following the way in format_iov. I've updated in new revision. And add macros for address translation.

zhngaj commented on prov/shm/src/smr_ep.c:
```diff
@@ -319,6 +319,88 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+	*map_name = calloc(1, sizeof(struct smr_ep_name));
+	if (!map_name) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL, "calloc error\n");
+		return -FI_ENOMEM;
```
Done.

zhngaj commented on prov/shm/src/smr_ep.c:
```diff
@@ -319,6 +319,88 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+	num = ofi_copy_from_iov(mapped_ptr, total_len, iov, count, 0);
+	if (num != total_len) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL, "copy from iov error\n");
+		goto err2;
```
Done.

zhngaj commented on prov/shm/src/smr_ep.c:
```diff
@@ -319,6 +319,88 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+err2:
+	dlist_remove(&(*map_name)->entry);
+	free(*map_name);
+	shm_unlink(shm_name);
```
Done.

zhngaj commented:
New revision (9ef862bca2153326c1ad576a0af0a25d47d472f9) which integrated all feedback so far has been pushed.

aingerson commented on prov/shm/src/smr_ep.c:
```diff
@@ -319,6 +319,86 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL, "copy from iov error\n");
+		ret = -FI_EIO;
+		goto munmap;
+	}
```
I don't love the name num for a size. Plus we can remove it entirely with
```if (ofi_copy_from_iov(mapped_ptr, total_len, iov, count, 0) != total_len)```

zhngaj replied:
Done.

aingerson commented on prov/shm/src/smr.h:
```diff
@@ -184,6 +185,9 @@ static inline const char *smr_no_prefix(const char *addr)
 			!(order & SMR_RMA_ORDER))
+#define smr_addr_offset(end, start) (uintptr_t)((char *) end - \
+				    (char *) start)
+#define smr_addr_end(start, offset) ((char *) start + (uintptr_t) offset)
```
Can we rename these parameter names and the second function name? It isn't clear to me with those names which parameter is which, specifically what "end" means.
Maybe
```
smr_get_offset(base, addr)
smr_get_addr(base, offset)
```

zhngaj replied:
Done.

aingerson commented on prov/shm/src/smr_msg.c:
```diff
@@ -185,26 +185,45 @@ static ssize_t smr_generic_sendmsg(struct smr_ep *ep, const struct iovec *iov,
+			 * mmap protocol
+			 * Once SAR protocol gets merged, we need set a threshold
+			 * for switching from SAR to mmap
+			 */
```
Can you remove the line mmap protocol line and turn this into a //TODO?

zhngaj replied:
Done.

aingerson commented on prov/shm/src/smr_ep.c:
```diff
@@ -319,6 +319,86 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+
+	msg_id = ep->msg_id++;
+	num = snprintf(shm_name, NAME_MAX - 1, "%s_%s_%ld", ep->name,
+		       (char *)peer_smr + peer_smr->name_offset, msg_id);
```
Why is the name "name1_name2_id"? Shouldn't just having the initiator peer name be enough?

zhngaj replied:
Yeah. Since we are using per endpoint msg_id instead of per peer msg_id, having the initiator peer name will suffice.

aingerson commented on prov/shm/src/smr_progress.c:
```diff
@@ -200,6 +207,75 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
+
+	num = snprintf(shm_name, NAME_MAX - 1, "%s_%s_%ld",
+		       ep->region->map->peers[peer_id].peer.name, ep->name,
+		       cmd->msg.hdr.msg_id);
```
I would make this a common inline so that if we change the format for the mmap file name, we only have it in one place

zhngaj replied:
Done.

aingerson commented on prov/shm/src/smr.h:
```diff
@@ -115,6 +115,7 @@ struct smr_tx_entry {
 	void		*context;
 	struct iovec	iov[SMR_IOV_LIMIT];
 	uint32_t	iov_count;
+	void		*map_ptr;
```
Shouldn't this be needed in the msg protocol as well? If so, it should probably be in the first patch. I also don't see any processing of the resp entry and cleaning up of the shm file in the previous commit. It should probably be there too.
We need to make sure we properly increment the cmd_cnt as well. I've added a fix and clarification for this in #5682 (adding a general progress_resp_entry function that always increments the cmd_cnt)

zhngaj replied:
In msg protocol, we don't need this map_ptr, as the sender does munmap/close as soon as it finishes copying the data into shm file. Upon receiving the signal from the receiver (resp->status==0), it just needs to shm_unlink and clean up map_name as follows.
```/c
if (pending->map_name) {
        shm_unlink(pending->map_name->name);
        dlist_remove(&pending->map_name->entry);
        free(pending->map_name);
}
```
map_ptr here is only for RMA read operation, so that the initiator does not need to get shm file name again and mmap again when progressing response.

> We need to make sure we properly increment the cmd_cnt as well. I've added a fix and clarification for this in #5682 (adding a general progress_resp_entry function that always increments the cmd_cnt)

Please see my reply in your comment `Again, see comment about looking at...`.

aingerson commented on prov/shm/src/smr_progress.c:
```diff
@@ -106,6 +106,15 @@ static void smr_progress_resp(struct smr_ep *ep)
+						pending->cmd.msg.hdr.size);
+				munmap(pending->map_ptr, pending->cmd.msg.hdr.size);
+			}
+
```
Again, see comment about looking at #5682 for generalization of smr_progress_resp to increment the cmd_cnt. I think we could pretty easily add a switch on the pending->cmd.msg.hdr.op to either just increment, copy from the map_ptr or copy from the tx buf depending on the op type.

zhngaj replied:
I saw your change in #5682 for generalization of smr_progress_resp to increment the cmd_cnt, and I agree with your idea of adding a switch. 

Regarding to the changes (formatting and generalization of smr_progress_resp), I prefer rebasing and updating once your PR #5682 gets merged, in case of you want to do more changes. Hope it's okay to you.

aingerson replied:
Yeah, that sounds good to me. Thank you!

aingerson commented on prov/shm/src/smr_rma.c:
```diff
@@ -152,27 +153,45 @@ ssize_t smr_generic_rma(struct smr_ep *ep, const struct iovec *iov,
+		smr_format_inject(cmd, smr_peer_addr(ep->region)[peer_id].addr,
+				  iov, iov_count, op, 0, data, op_flags,
+				  peer_smr, tx_buf);
+	} else {
```
This switching of the protocol order is unrelated to this patch. Please separate.

zhngaj replied:
Done.

zhngaj commented on prov/util/src/util_shm.c:
```diff
@@ -232,6 +254,13 @@ void smr_map_to_endpoint(struct smr_region *region, int index)
+	/*
+	 * Once peer's region has been mapped succesully,
+	 * the map->peers[index].peer.addr wont't be FI_ADDR_UNSPEC
+	 */
```
Done.

zhngaj commented on include/ofi_shm.h:
```diff
@@ -80,6 +80,11 @@ enum {
+#define SMR_CMA_CAP_MISS	(1 << 0)
+#define SMR_CMA_CAP_ON		(1 << 1)
+#define SMR_CMA_CAP_OFF		(1 << 2)
+
```
Yes, we can. Done.

zhngaj commented on prov/shm/src/smr_init.c:
```diff
@@ -140,12 +140,13 @@ static int smr_getinfo(uint32_t version, const char *node, const char *service,
 			cur->ep_attr->max_order_war_size = 0;
 		}
+		if (mr_mode & FI_MR_VIRT_ADDR)
+			cur->domain_attr->mr_mode = FI_MR_VIRT_ADDR;
```
My understanding is that if app requires FI_MR_VIRT_ADDR via hints, and shm provider supports FI_MR_VIRT_ADDR, the returned info should have FI_MR_VIRT_ADDR bit, regardless of CMA capability. 
Without this change, FI_MR_VIRT_ADDR bit will be cleared after util_getinfo call. If CMA is not available, fast_rma will be disabled, and FI_MR_VIRT_ADDR won't be set.

zhngaj commented on prov/shm/src/smr_ep.c:
```diff
@@ -319,6 +319,86 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+		return -errno;
+	}
+
+	map_name = calloc(1, sizeof(struct smr_ep_name));
```
Done.

shefty commented:
> > +		if (mr_mode & FI_MR_VIRT_ADDR)
> +			cur->domain_attr->mr_mode = FI_MR_VIRT_ADDR;
> 
> My understanding is that if app requires FI_MR_VIRT_ADDR via hints, and shm provider
> supports FI_MR_VIRT_ADDR, the returned info should have FI_MR_VIRT_ADDR bit, regardless
> of CMA capability.

FI_MR_VIRT_ADDR is a requirement *from the provider*, not the app.  The app indicates if it can support that requirement.  The provider toggles the bit on/off if it is needed.

shefty commented:
> Sorry for the confusion. Let me put my question in this way:
> suppose that app puts FI_MR_VIRT_ADDR bit in hints, and the app calls fi_getinfo.
> Is FI_MR_VIRT_ADDR bit supposed to be returned from shm provider?

It is not required to be returned.

> The current logic (without my change) is:
> 
> *	if CMA available, rma fast path will be enabled, FI_MR_VIRT_ADDR will be set and
> returned.
> *	if no CMA, rma fast path will be disabled, thus NO FI_MR_VIRT_ADDR bit will be
> set and retuned.
> 	Is such behavior expected?

Yes, this is acceptable.  FI_MR_VIRT_ADDR is less scalable than the alternative (0-based offset).  The goal of the provider should be to disable mr_mode bits, unless they are required for performance reasons or are a hard requirement.

Btw, for some reason I cannot load this PR on github without it timing out.  So I have to respond via email.

zhngaj commented:
> Yes, this is acceptable. FI_MR_VIRT_ADDR is less scalable than the alternative (0-based offset). The goal of the provider should be to disable mr_mode bits, unless they are required for performance reasons or are a hard requirement. Btw, for some reason I cannot load this PR on github without it timing out. So I have to respond via email.

Thank you very much for the explanation. The returned mr_mode is inconsistent between different OSes (one with CMA vs. one without CMA) though, which might be confusing for other users.  

> Btw, for some reason I cannot load this PR on github without it timing out.  So I have to respond via email

I have same loading issue, not sure why.

wzamazon commented:
@shefy regarding the FI_MR_VIRT_ADDR issue. I understand that FI_MR_VIRT_ADDR is less desirable, but considering that CMA is using it, maybe it is better for mmap to use it for consistency? Otherwise, application would see different mr_mode on different OS, and will need to introduce a branch to check mr_mode for each RMA operation.

shefty commented:
VIRT_ADDR requires exchanging the base address with the peer.  The app can still provide 0 as the base address if it doesn't want to optimize for it and avoid the message exchange completely.  Since the address exchange comes after a memory registration call, I'm not concerned about an extra if statement in the code.

If shm is being used in conjunction by another provider without direct knowledge of the app, then I agree we can look at aligning all the mr_mode bits.  My guess is that's like to just move the checks somewhere else, and not eliminate them.  The goal is to support apps that have this bit set and unset.

zhngaj commented:
Rebased and updated the mmap code with the recently merged PR #5762.

wzamazon commented on prov/efa/src/rxr/rxr.h:
```diff
@@ -189,7 +189,6 @@ extern struct util_prov rxr_util_prov;
 extern struct efa_ep_addr *local_efa_addr;
 
 struct rxr_env {
-	int shm_cma_cap;
```
shm_cma_cap should not have been introduced at all.

zhngaj replied:
We introduced this env. variable to check if we need mr registration in the last commit.

wzamazon replied:
Is it used in last commit? If not, maybe we remove it from last commit?

zhngaj replied:
It is in the last commit. 
```/c
} else if (rxr_env.enable_shm_transfer && peer->is_local && rxr_env.shm_cma_cap) {
        for (i = 0; i < tx_entry->iov_count; ++i)
                read_iov[i].key = 0;
```

wzamazon commented on prov/efa/src/rxr/rxr_atomic.c:
```diff
@@ -38,6 +38,16 @@
+
+	if (!(shm_info->domain_attr->mr_mode & FI_MR_VIRT_ADDR)) {
+		for (i = 0; i < msg->rma_iov_count; i++)
+			*(uint64_t *)&msg->rma_iov[i].addr = 0;
```
This line is quite confusing, should it just be ```msg->rma_iov[I].addr = 0```?

zhngaj replied:
This is for accessing the constant member `const struct fi_rma_iov *rma_iov` in struct  fi_msg_rma.

wzamazon replied:
If it is defined as constant, we should not forcefully change it. That might result in unexpected behavior from application. The application might reuse the rma_iov later and did not know the value was changed.

zhngaj replied:
The struct fi_msg_rma msg here in what we allocate internally, not the application.

wzamazon replied:
I know msg is internally allocated. It is ```shm_msg``` in the caller.  but ```shm_msg``` is a mem copy of ```msg```. Note ```msg->rma_iov``` is a pointer, not an array, See the following definition:

```
struct fi_msg_atomic {
	const struct fi_ioc *msg_iov; /* local scatter-gather array */
	void                **desc;   /* local access descriptors */
	size_t              iov_count;/* # elements in ioc */
	const void          *addr;    /* optional endpoint address */
	const struct fi_rma_ioc *rma_iov; /* remote SGL */
	size_t              rma_iov_count;/* # elements in remote SGL */
	enum fi_datatype    datatype; /* operand datatype */
	enum fi_op          op;       /* atomic operation */
	void                *context; /* user-defined context */
	uint64_t            data;     /* optional data */
};

```

so ```shm_msg->rma_iov``` and ```msg->rma_iov``` pointing to the same array that is allocated by caller, who does not expect it to be changed.

zhngaj replied:
I see your point. Updated accordingly.

aingerson commented on prov/shm/src/smr_ep.c:
```diff
@@ -298,6 +298,80 @@ void smr_format_iov(struct smr_cmd *cmd, const struct iovec *iov, size_t count,
+		ret = -FI_ENOMEM;
+		goto unlink_close;
+	}
+	strncpy((map_name)->name, shm_name, NAME_MAX - 1);
```
We can save a copy and simplify the error case for this is we allocate map_name first and then call mmap_shm_name with map_name->name

zhngaj replied:
Yes, thank you. Updated.

aingerson commented on prov/shm/src/smr.h:
```diff
@@ -206,6 +211,13 @@ struct smr_ep {
 #define smr_ep_rx_flags(smr_ep) ((smr_ep)->util_ep.rx_op_flags)
 #define smr_ep_tx_flags(smr_ep) ((smr_ep)->util_ep.tx_op_flags)
 
+static inline int mmap_shm_name(char *shm_name, const char *ep_name,
```
Rename to smr_mmap_name for prefix consistency

zhngaj replied:
Done.

aingerson commented on prov/shm/src/smr.h:
```diff
@@ -184,6 +185,9 @@ static inline const char *smr_no_prefix(const char *addr)
 			!(order & SMR_RMA_ORDER))
+#define smr_get_offset(base, addr) (uintptr_t)((char *) addr - \
+				    (char *) base)
+#define smr_get_addr(base, offset) ((char *) base + (uintptr_t) offset)
```
Can you separate this into its own patch and update the other instances of this? (format_inject, format_iov, progress_inject, progress_iov, progress_resp_entry, etc)

zhngaj replied:
Done.

aingerson commented on prov/shm/src/smr_rma.c:
```diff
@@ -157,8 +157,14 @@ ssize_t smr_generic_rma(struct smr_ep *ep, const struct iovec *iov,
+	    !(op_flags & FI_DELIVERY_COMPLETE)) {
+		smr_format_inline(cmd, iov, iov_count);
+	} else if (total_len <= SMR_INJECT_SIZE && op != ofi_op_read_req &&
+		   !(op_flags & FI_DELIVERY_COMPLETE)) {
```
I understand and support the purpose of these changes but these checks are starting to get a little crazy. Can we simplify this in any way?

zhngaj replied:
I improved a little bit, but still not perfect.

aingerson commented:
@zhngaj btw #5682 was merged so this can be updated to include the resp_entry changes

zhngaj commented:
> be updated to include the resp_entry changes

@aingerson Thanks for the reminder. Sorry for not able to review your PR #5682 one more time. When doing the rebase, it seems that we are using inject protocol rather than inline protocol for small size RMA write. Is it intended or am I missing something?

aingerson commented:
> > be updated to include the resp_entry changes
> 
> @aingerson Thanks for the reminder. Sorry for not able to review your PR #5682 one more time. When doing the rebase, it seems that we are using inject protocol rather than inline protocol for small size RMA write. Is it intended or am I missing something?

Hmm yes you're right. I didn't catch that. Could you add a patch in this series that changes the inject check to 
```
else if (total_len > SMR_MSG_DATA_LEN || (smr_env.disable_cma && op == ofi_op_read_req))
```
Thank you!

zhngaj commented:
> Hmm yes you're right. I didn't catch that. Could you add a patch in this series that changes the inject check to

Sure.

aingerson commented on prov/shm/src/smr_rma.c:
```diff
@@ -156,20 +157,33 @@ ssize_t smr_generic_rma(struct smr_ep *ep, const struct iovec *iov,
-	if (!smr_env.disable_cma && (total_len > SMR_INJECT_SIZE ||
-	    op != ofi_op_write || op_flags & FI_DELIVERY_COMPLETE)) {
+	if (total_len > SMR_INJECT_SIZE || (!smr_env.disable_cma &&
+	    (op != ofi_op_write || op_flags & FI_DELIVERY_COMPLETE))) {
```
I don't think the reorganization helps here and I'm not sure it's correct. I can't tell. 

I think we also need to assess whether we still want to have the DISABLE_CMA variable and, if so, what effect it's going to have. I think if we do keep it, we need to move it out of the protocol and into the cma_check where we force cma_cap to OFF if disable_cma is on. But I don't really think the environment variable is very useful here anymore. It was mostly a workaround until we had an alternate large message protocol/proper CMA detection. Right now, I think we will have an issue here because if we disable cma with the environment variable, this check will fail and we will go into the inject call but then the message will be too large. I think we should remove it entirely.

zhngaj replied:
Yeah, it makes sense, as we already have a peer-to-peer cma detection.
I would remove the env. variable, smr_check_ptrace_scope() entirely. 
Thank you.

zhngaj replied:
The change is in commit 32b08cb

zhngaj commented on include/ofi_shm.h:
```diff
@@ -80,6 +80,13 @@ enum {
 
+/* CMA capability */
+enum {
+	SMR_CMA_CAP_NA,
```
This is suggested by Alexia, which I also feel is better.

zhngaj commented on prov/efa/src/rxr/rxr_atomic.c:
```diff
@@ -38,6 +38,16 @@
+{
+	int i;
+
+	if (!(shm_info->domain_attr->mr_mode & FI_MR_VIRT_ADDR)) {
```
Yeah. Thank you.

zhngaj commented:
AWS CI failure is due to CI system related issue

zhngaj commented:
While two CI (appveyor, travis-ci) are still pending, I don't expect them to fail. 
Could you @aingerson @ofiwg/aws-efa-team please give the PR one more pass, so that we can merge it ASAP.

aingerson commented on prov/shm/src/smr_rma.c:
```diff
@@ -157,8 +157,30 @@ ssize_t smr_generic_rma(struct smr_ep *ep, const struct iovec *iov,
+		smr_format_inline(cmd, iov, iov_count);
+	} else if (total_len <= SMR_INJECT_SIZE &&
+		   (smr_env.disable_cma || op == ofi_op_write) &&
+		   !(op_flags & FI_DELIVERY_COMPLETE)) {
```
The check for CMA and op were put there because of the lack of fallback protocols. 
Now it can just be:
```
else if (total_len <= SMR_INJECT_SIZE && !(op_flags & FI_DELIVERY))
```

zhngaj replied:
Hmm. I thought we still want to use iov protocol for read if CMA is available.
With your suggested change, do you mean we prefer inject protocol instead of iov protocol for medium/small RMA read even though CMA is available?

zhngaj replied:
Updated with your suggested change. Now
- inline  -  small size write
- inject  -  small size read, medium size read/write
- iov      -   large size read/write (w. CMA)
- mmap -  large size read/write (w/o CMA)

aingerson commented on prov/shm/src/smr_rma.c:
```diff
@@ -161,7 +161,7 @@ ssize_t smr_generic_rma(struct smr_ep *ep, const struct iovec *iov,
 		smr_format_inline(cmd, iov, iov_count);
 	} else if (total_len <= SMR_INJECT_SIZE &&
-		   (smr_env.disable_cma || op == ofi_op_write) &&
+		   (ep->region->cma_cap == SMR_CMA_CAP_OFF || op == ofi_op_write) &&
```
The cap check shouldn't be needed here because CMA doesn't affect whether we take the inject path or not (should be removed in the previous commit)

zhngaj replied:
Updated.

wzamazon commented:
Thank you! I am satisfied with EFA related patches

zhngaj commented:
Thank you @shefty @aingerson @wzamazon @rajachan very much, really appreciate your feedback.
I will go ahead and merge.
